### PR TITLE
libnet_if_addr.c: fix segfault when number of IPs > 512

### PR DESCRIPTION
**Pull request by aroulin**
The current code has a hardcoded limit of 512 IP addresses. The static array will be overflowed when more addresses are present.

Fix the static limit by using realloc when the array needs to grow to accomodate more interfaces.

I verified on a setup with 1200+ interfaces/IPs and also ran valgrind to confirm there's no memory leak.

Closes: #149

### CODE REVIEW

File: src/libnet_if_addr.c [hunk 1]
```diff
@@ -46,7 +46,9 @@
 
 #include "../include/ifaddrlist.h"
 
-#define MAX_IPADDR 512
+#define MAX_IP_ADDR 512
+#define IPADDR_INITIAL_BUFSIZE MAX_IP_ADDR
+#define IPADDR_GROW_FACTOR 1.5
```
Comment (troglobit) [inline src/libnet_if_addr.c:52]: Please change this to not use floating point.  I'd rather see `IP_ADDR_GROW 256` or something.

To track number of current addresses, use a global (to this file) construct:

```c
static size_t ip_addr_num = MAX_IP_ADDR;
```

File: src/libnet_if_addr.c [hunk 2]
```diff
@@ -104,10 +106,20 @@ int
...
     (void)dev; /* unused */
-    static struct libnet_ifaddr_list ifaddrlist[MAX_IPADDR];
+    static struct libnet_ifaddr_list *ifaddrlist = NULL, *ifaddrlist_tmp = NULL;
+    static size_t ifaddrlist_len = 0;
     struct ifaddrs *ifap, *ifa;
     int i = 0;
-    memset (ifaddrlist, 0 , sizeof(ifaddrlist));
+
+    if (!ifaddrlist) {
+        ifaddrlist = calloc(IPADDR_INITIAL_BUFSIZE, sizeof(struct libnet_ifaddr_list));
```
Comment (troglobit) [inline src/libnet_if_addr.c:115]: This then becomes

```c
ifaddrlist = calloc(ip_addr_num, sizeof(struct libnet_ifaddr_list));
```

File: src/libnet_if_addr.c [hunk 3]
```diff
@@ -130,6 +142,18 @@ libnet_ifaddrlist(register struct libnet_ifaddr_list **ipaddrp, char *dev, regis
             ifaddrlist[i].addr = ((struct sockaddr_in *)ifa->ifa_addr)->sin_addr.s_addr;
             ++i;
         }
+
+        if (i == ifaddrlist_len) {
+            ifaddrlist_tmp = realloc(ifaddrlist, ifaddrlist_len *
+				     IPADDR_GROW_FACTOR * sizeof(struct libnet_ifaddr_list));
```
Comment (troglobit) [inline src/libnet_if_addr.c:148]: and this becomes:

```C
ip_addr_num += IP_ADDR_GROW;
ifaddrlist_tmp = realloc(ifaddrlist, ip_addr_num * sizeof(struct libnet_ifaddr_list));
```

which is also a lot easier to read.

**Review by troglobit — CHANGES_REQUESTED** [review]

**Comment by troglobit** [comment]
Thanks for submitting this PR! Please update it with my comments and rebase+squash that.

File: src/libnet_if_addr.c (hunk 1 above)
Comment (aroulin) [reply src/libnet_if_addr.c:52]: The idea of the floating point growth factor is to scale better to a large number of interfaces.

Say we have 3000 interfaces/IP addresses:

With static 256 growth factor:
`(3000 - 512)/256 = 9.7`
we will need already 10 realloc

with the multiply 1.5 growth factor:
`512 * (1.5)ˆx >= 3000`
`x >= 4.5`
Only need 5 reallocs

1.5 was chosen as a good ratio for dynamic array growth and previous left memory holes reuse (https://stackoverflow.com/questions/1100311/what-is-the-ideal-growth-rate-for-a-dynamically-allocated-array)

Let me know what you think and if you'd rather keep the constant addition. I will definitely make the global `ip_addr_num` change in both cases.

Thank you very much for the review

Comment (troglobit) [reply src/libnet_if_addr.c:52]: OK that's a good motivation, I appreciate your explanation.  I basically just dislike the use of floating point, so if you can address that I'm happy. Eg num += num >> 2; or something.

Comment (aroulin) [reply src/libnet_if_addr.c:52]: That's a good idea, thank you! I also felt an umconfortable shiver in my body when I started using floating points. Never know when NaN or 0.000001 is going to creep up on you.

**Comment by troglobit** [comment]
Looks awesome, great work! Unleashing CI, when it completes I'll merge :-)

**Comment by troglobit** [comment]
Unfortunately this PR caused a few regressions.  I dusted off my old MacBook Pro to figure out was going on.  Commits f7f0a11 (macOS and "other" OSs) and 2bc784f should fix the regression, although the fix for Windows has *not* been tested due to lack of a Windows machine.

**Pull request merged**